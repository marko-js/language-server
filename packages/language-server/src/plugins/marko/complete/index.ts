import { NodeType } from "@marko/language-tools";
import { CompletionItem } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { AttrName } from "./AttrName";
import { Import } from "./Import";
import { OpenTagName } from "./OpenTagName";
import { Tag } from "./Tag";

export function provideCompletions(
  doc: MarkoVirtualCode,
  offset: number,
): CompletionItem[] | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.AttrName:
      return AttrName(node, doc, offset);
    case NodeType.Import:
    case NodeType.Static:
      return Import(node, doc);
    case NodeType.Tag:
      return Tag(node, doc, offset);
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
