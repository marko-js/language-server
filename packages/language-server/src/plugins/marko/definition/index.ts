import { NodeType } from "@marko/language-tools";
import { LocationLink } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { AttrName } from "./AttrName";
import { OpenTagName } from "./OpenTagName";

export function provideDefinitions(
  doc: MarkoVirtualCode,
  offset: number,
): LocationLink[] | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.AttrName:
      return AttrName(node, doc);
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
