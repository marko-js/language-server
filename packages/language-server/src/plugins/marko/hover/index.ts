import { NodeType } from "@marko/language-tools";
import { Hover } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { OpenTagName } from "./OpenTagName";

export function provideHover(
  doc: MarkoVirtualCode,
  offset: number,
): Hover | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
