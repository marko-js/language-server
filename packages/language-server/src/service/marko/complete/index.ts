import type { CompletionItem, CompletionParams } from "vscode-languageserver";
import { NodeType } from "@marko/language-tools";

import { MarkoFile, getMarkoFile } from "../../../utils/file";
import type { Plugin, Result } from "../../types";

import { AttrName } from "./AttrName";
import { AttrValue } from "./AttrValue";
import { Import } from "./Import";
import { OpenTagName } from "./OpenTagName";
import { Tag } from "./Tag";

export type CompletionResult = Result<CompletionItem[]>;
export interface CompletionMeta<N = unknown> {
  file: MarkoFile;
  params: CompletionParams;
  offset: number;
  node: N;
}

const handlers: Record<
  string,
  (data: CompletionMeta<any>) => CompletionResult
> = {
  Tag,
  OpenTagName,
  AttrName,
  AttrValue,
  Import,
};

export const doComplete: Plugin["doComplete"] = async (doc, params) => {
  const file = getMarkoFile(doc);
  const offset = doc.offsetAt(params.position);
  const node = file.parsed.nodeAt(offset);
  return {
    items:
      (await handlers[NodeType[node.type]]?.({
        file,
        params,
        offset,
        node,
      })) || [],
    isIncomplete: true,
  };
};
