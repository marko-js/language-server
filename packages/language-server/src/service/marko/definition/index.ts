import type { DefinitionLink, DefinitionParams } from "vscode-languageserver";
import { NodeType } from "@marko/language-tools";

import { MarkoFile, getMarkoFile } from "../../../utils/file";
import type { Plugin, Result } from "../../types";

import { AttrName } from "./AttrName";
import { OpenTagName } from "./OpenTagName";

export type DefinitionResult = Result<DefinitionLink[]>;
export interface DefinitionMeta<N = unknown> {
  file: MarkoFile;
  params: DefinitionParams;
  offset: number;
  node: N;
}

const handlers: Record<
  string,
  (data: DefinitionMeta<any>) => DefinitionResult
> = {
  OpenTagName,
  AttrName,
};

export const findDefinition: Plugin["findDefinition"] = async (doc, params) => {
  const file = getMarkoFile(doc);
  const offset = doc.offsetAt(params.position);
  const node = file.parsed.nodeAt(offset);
  return (
    (await handlers[NodeType[node.type]]?.({
      file,
      params,
      offset,
      node,
    })) || []
  );
};
