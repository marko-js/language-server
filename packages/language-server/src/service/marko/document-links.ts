import type { DocumentLink } from "vscode-languageserver";
import { type Node, NodeType } from "@marko/language-tools";
import { MarkoFile, processDoc } from "../../utils/file";

import resolveUrl from "../../utils/resolve-url";
import type { Plugin } from "../types";

import isDocumentLinkAttr from "./util/is-document-link-attr";

const importTagReg = /(['"])<((?:[^\1\\>]+|\\.)*)>?\1/g;

export const findDocumentLinks: Plugin["findDocumentLinks"] = async (doc) => {
  return processDoc(doc, extractDocumentLinks);
};

/**
 * Iterate over the Marko CST and extract all the file links in the document.
 */
function extractDocumentLinks({
  uri,
  scheme,
  parsed,
  code,
  project: { lookup },
}: MarkoFile): DocumentLink[] {
  if (scheme !== "file") {
    return [];
  }

  const links: DocumentLink[] = [];
  const { program, read } = parsed;
  const visit = (node: Node.ChildNode) => {
    switch (node.type) {
      case NodeType.AttrTag:
        if (node.body) {
          for (const child of node.body) {
            visit(child);
          }
        }
        break;
      case NodeType.Tag:
        if (node.attrs && node.nameText) {
          for (const attr of node.attrs) {
            if (isDocumentLinkAttr(code, node, attr)) {
              const resolved = resolveUrl(
                read(attr.value.value).slice(1, -1),
                uri
              );
              if (resolved) {
                links.push({
                  range: parsed.locationAt(attr.value.value),
                  target: resolveUrl(read(attr.value.value).slice(1, -1), uri),
                });
              }
            }
          }
        }
        if (node.body) {
          for (const child of node.body) {
            visit(child);
          }
        }

        break;
    }
  };

  for (const node of program.static) {
    // check for import statement (this currently only support the tag import shorthand).
    if (node.type === NodeType.Import) {
      importTagReg.lastIndex = 0;
      const value = parsed.read(node);
      const match = importTagReg.exec(value);
      if (match) {
        const [{ length }, , tagName] = match;
        const tagDef = lookup.getTag(tagName);
        const fileForTag = tagDef && (tagDef.template || tagDef.renderer);

        if (fileForTag) {
          links.push({
            range: parsed.locationAt({
              start: node.start + match.index,
              end: node.start + match.index + length,
            }),
            target: fileForTag,
          });
        }
      }
    }
  }

  for (const node of program.body) {
    visit(node);
  }

  return links;
}
