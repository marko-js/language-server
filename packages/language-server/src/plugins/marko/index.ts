import {
  LanguageServicePlugin,
  LanguageServicePluginInstance,
} from "@volar/language-service";
import { TextDocument } from "vscode-languageserver-textdocument";
import { URI } from "vscode-uri";

import { MarkoVirtualCode } from "../../language";
import { provideCompletions } from "./complete";
import { provideDefinitions } from "./definition";
import { provideHover } from "./hover";
import { provideValidations } from "./validate";
// import { provideDocumentSymbols } from "./document-symbols";

export const create = (
  _: typeof import("typescript"),
): LanguageServicePlugin => {
  return {
    name: "marko",
    capabilities: {
      hoverProvider: true,
      definitionProvider: true,
      diagnosticProvider: {
        interFileDependencies: false,
        workspaceDiagnostics: false,
      },
      completionProvider: {
        triggerCharacters: [
          ".",
          ":",
          "<",
          ">",
          "@",
          "/",
          '"',
          "'",
          "`",
          " ",
          "=",
          "*",
          "#",
          "$",
          "+",
          "^",
          "(",
          "[",
          "-",
        ],
      },
    },
    create(context): LanguageServicePluginInstance {
      return {
        // TODO: Is this necessary?
        // provideDocumentSymbols(document, token) {
        //   if (token.isCancellationRequested) return;
        //   return worker(document, (virtualCode) => {
        //     return provideDocumentSymbols(virtualCode);
        //   });
        // },
        provideDefinition(document, position, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            return provideDefinitions(virtualCode, offset);
          });
        },
        provideDiagnostics(document, token) {
          if (token.isCancellationRequested) return;
          return worker(document, async (virtualCode) => {
            return await provideValidations(virtualCode);
          });
        },
        provideHover(document, position, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            return provideHover(virtualCode, offset);
          });
        },
        provideCompletionItems(document, position, _, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            const completions = provideCompletions(virtualCode, offset);

            if (completions) {
              return {
                isIncomplete: false,
                items: completions.map((it) => {
                  it.data = { source: "marko" };
                  return it;
                }),
              };
            }

            return {
              items: [],
              isIncomplete: true,
            };
          });
        },
      };

      function worker<T>(
        document: TextDocument,
        callback: (markoDocument: MarkoVirtualCode) => T,
      ): T | undefined {
        const decoded = context.decodeEmbeddedDocumentUri(
          URI.parse(document.uri),
        );
        const sourceScript =
          decoded && context.language.scripts.get(decoded[0]);
        const virtualCode =
          decoded && sourceScript?.generated?.embeddedCodes.get(decoded[1]);
        if (!(virtualCode instanceof MarkoVirtualCode)) return;

        return callback(virtualCode);
      }
    },
  };
};
