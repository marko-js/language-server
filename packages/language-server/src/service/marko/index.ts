import type { Plugin } from "../types";

import { doComplete } from "./complete";
import { doValidate } from "./validate";
import { doHover } from "./hover";
import { findDefinition } from "./definition";
import { findDocumentLinks } from "./document-links";
import { findDocumentSymbols } from "./document-symbols";
import { format } from "./format";

export default {
  doComplete,
  doValidate,
  doHover,
  findDefinition,
  findDocumentLinks,
  findDocumentSymbols,
  format,
} as Plugin;
