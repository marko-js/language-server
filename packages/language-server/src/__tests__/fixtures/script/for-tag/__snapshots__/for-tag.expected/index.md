## Hovers
### Ln 14, Col 5
```marko
  12 |
  13 | <for|item, index, all| of=list>
> 14 |   ${item} ${index} ${all}
     |     ^ (parameter) item: {
    readonly value: 1;
} | {
    readonly value: 2;
} | {
    readonly value: 3;
}
  15 | //  ^?      ^?       ^?
  16 | </for>
  17 |
```

### Ln 14, Col 13
```marko
  12 |
  13 | <for|item, index, all| of=list>
> 14 |   ${item} ${index} ${all}
     |             ^ (parameter) index: number
  15 | //  ^?      ^?       ^?
  16 | </for>
  17 |
```

### Ln 14, Col 22
```marko
  12 |
  13 | <for|item, index, all| of=list>
> 14 |   ${item} ${index} ${all}
     |                      ^ (parameter) all: readonly [{
    readonly value: 1;
}, {
    readonly value: 2;
}, {
    readonly value: 3;
}]
  15 | //  ^?      ^?       ^?
  16 | </for>
  17 |
```

### Ln 18, Col 18
```marko
  16 | </for>
  17 |
> 18 | <for of=list by=(item, index) => `${item}-${index}`>
     |                  ^ (parameter) item: {
    readonly value: 1;
} | {
    readonly value: 2;
} | {
    readonly value: 3;
}
  19 | //               ^?    ^?
  20 | </for>
  21 |
```

### Ln 18, Col 24
```marko
  16 | </for>
  17 |
> 18 | <for of=list by=(item, index) => `${item}-${index}`>
     |                        ^ (parameter) index: number
  19 | //               ^?    ^?
  20 | </for>
  21 |
```

### Ln 29, Col 3
```marko
  27 |
  28 | <effect() {
> 29 |   hoistedFromForOf;
     |   ^ const hoistedFromForOf: 1 | 2 | 3
  30 | //^?
  31 | }/>
  32 |
```

### Ln 36, Col 5
```marko
  34 |
  35 | <for|key, value| in=record>
> 36 |   ${key} ${value}
     |     ^ (parameter) key: "a" | "b"
  37 | //  ^?     ^?
  38 | </for>
  39 |
```

### Ln 36, Col 12
```marko
  34 |
  35 | <for|key, value| in=record>
> 36 |   ${key} ${value}
     |            ^ (parameter) value: 1 | 2
  37 | //  ^?     ^?
  38 | </for>
  39 |
```

### Ln 40, Col 20
```marko
  38 | </for>
  39 |
> 40 | <for in=record by=(value, key) => `${value}-${key}`>
     |                    ^ (parameter) value: 1 | 2
  41 | //                 ^?     ^?
  42 | </for>
  43 |
```

### Ln 40, Col 27
```marko
  38 | </for>
  39 |
> 40 | <for in=record by=(value, key) => `${value}-${key}`>
     |                           ^ (parameter) key: "a" | "b"
  41 | //                 ^?     ^?
  42 | </for>
  43 |
```

### Ln 49, Col 3
```marko
  47 |
  48 | <effect() {
> 49 |   hoistedFromForIn;
     |   ^ const hoistedFromForIn: "a" | "b"
  50 | //^?
  51 | }/>
  52 |
```

### Ln 54, Col 5
```marko
  52 |
  53 | <for|index| to=10>
> 54 |   ${index}
     |     ^ (parameter) index: number
  55 | //  ^?
  56 | </for>
  57 |
```

### Ln 58, Col 16
```marko
  56 | </for>
  57 |
> 58 | <for to=10 by=(index) => `${index}`>
     |                ^ (parameter) index: number
  59 | //             ^?
  60 | </for>
  61 |
```

### Ln 63, Col 5
```marko
  61 |
  62 | <for|index| from=1 to=10>
> 63 |   ${index}
     |     ^ (parameter) index: number
  64 | //  ^?
  65 | </for>
  66 |
```

### Ln 68, Col 5
```marko
  66 |
  67 | <for|index| to=10 step=2>
> 68 |   ${index}
     |     ^ (parameter) index: number
  69 | //  ^?
  70 | </for>
  71 |
```

### Ln 77, Col 3
```marko
  75 |
  76 | <effect() {
> 77 |   hoistedFromForTo;
     |   ^ const hoistedFromForTo: number
  78 | //^?
  79 | }/>
  80 |
```

## Diagnostics
### Ln 26, Col 6
```marko
  24 | </for>
  25 |
> 26 | <for|item| of=list></for>
     |      ^^^^ 'item' is declared but its value is never read.
  27 |
  28 | <effect() {
  29 |   hoistedFromForOf;
```

### Ln 28, Col 2
```marko
  26 | <for|item| of=list></for>
  27 |
> 28 | <effect() {
     |  ^^^^^^ Cannot find name 'effect'.
  29 |   hoistedFromForOf;
  30 | //^?
  31 | }/>
```

### Ln 48, Col 2
```marko
  46 | </for>
  47 |
> 48 | <effect() {
     |  ^^^^^^ Cannot find name 'effect'.
  49 |   hoistedFromForIn;
  50 | //^?
  51 | }/>
```

### Ln 76, Col 2
```marko
  74 | </for>
  75 |
> 76 | <effect() {
     |  ^^^^^^ Cannot find name 'effect'.
  77 |   hoistedFromForTo;
  78 | //^?
  79 | }/>
```

### Ln 81, Col 2
```marko
  79 | }/>
  80 |
> 81 | <for|index|>
     |  ^^^^^^^^^ No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ renderBody: (index: any) => MarkoReturn<void>; }' is not assignable to parameter of type '({ from?: number | undefined; to: number; step?: number | undefined; } | { in: unknown; } | { of: readonly unknown[] | Iterable<unknown>; }) & { renderBody?: AnyMarkoBody | undefined; by?: ((...args: unknown[]) => string) | undefined; }'.
      Type '{ renderBody: (index: any) => MarkoReturn<void>; }' is not assignable to type '{ of: readonly unknown[] | Iterable<unknown>; } & { renderBody?: AnyMarkoBody | undefined; by?: ((...args: unknown[]) => string) | undefined; }'.
        Property 'of' is missing in type '{ renderBody: (index: any) => MarkoReturn<void>; }' but required in type '{ of: readonly unknown[] | Iterable<unknown>; }'.
  82 |   Should error
  83 | </for>
  84 |
```

### Ln 81, Col 6
```marko
  79 | }/>
  80 |
> 81 | <for|index|>
     |      ^^^^^ 'index' is declared but its value is never read.
  82 |   Should error
  83 | </for>
  84 |
```

