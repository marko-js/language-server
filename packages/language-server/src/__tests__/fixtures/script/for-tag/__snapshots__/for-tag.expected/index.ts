export interface Input {}
abstract class Component extends Marko.Component<Input> {}
export { type Component };
(function (this: void) {
  const input = Marko._.any as Input;
  const component = Marko._.any as Component;
  const state = Marko._.state(component);
  const out = Marko._.out;
  const $signal = Marko._.any as AbortSignal;
  const $global = Marko._.getGlobal(
    // @ts-expect-error We expect the compiler to error because we are checking if the MarkoRun.Context is defined.
    (Marko._.error, Marko._.any as MarkoRun.Context),
  );
  Marko._.noop({ component, state, out, input, $global, $signal });
  const __marko_internal_tag_1 = Marko._.resolveTemplate(
    import("../../../components/let/index.marko"),
  );
  Marko._.assertRendered(
    Marko._.rendered,
    1,
    Marko._.renderTemplate(__marko_internal_tag_1)()()({
      value: [
        {
          value: 1,
        },
        {
          value: 2,
        },
        {
          value: 3,
        },
      ] as const,
    }),
  );
  const list = Marko._.rendered.returns[1].value;
  Marko._.forOfTag(
    {
      /*for*/ of: list,
    },
    () => {
      return Marko._.voidReturn;
    },
  );
  Marko._.forOfTag(
    {
      /*for*/ of: list,
    },
    (item, index, all) => {
      item;
      index;
      all;
      return Marko._.voidReturn;
    },
  );
  Marko._.forOfTag(
    {
      /*for*/ of: list,
      by: (item, index) => `${item}-${index}`,
    },
    () =>
      //  ^?      ^?       ^?

      {
        return Marko._.voidReturn;
      },
  );
  Marko._.assertRendered(
    Marko._.rendered,
    2,
    Marko._.forOfTag(
      {
        /*for*/ of: list,
      },
      (
        //               ^?    ^?
        item,
      ) => {
        const __marko_internal_tag_2 = Marko._.resolveTemplate(
          import("../../../components/const/index.marko"),
        );
        Marko._.assertRendered(
          Marko._.rendered,
          3,
          Marko._.renderTemplate(__marko_internal_tag_2)()()({
            value: item,
          }),
        );
        const { value: hoistedFromForOf } = Marko._.rendered.returns[3].value;
        return new (class MarkoReturn<Return = void> {
          [Marko._.scope] = { hoistedFromForOf };
          declare return: Return;
          constructor(_?: Return) {}
        })();
      },
    ),
  );
  Marko._.forOfTag(
    {
      /*for*/ of: list,
    },
    (item) => {
      return Marko._.voidReturn;
    },
  );
  const __marko_internal_tag_3 = Marko._.interpolated`effect`;
  Marko._.renderDynamicTag(__marko_internal_tag_3)()()({
    value() {
      hoistedFromForOf;
      //^?
    },
  });
  const __marko_internal_tag_4 = Marko._.resolveTemplate(
    import("../../../components/let/index.marko"),
  );
  Marko._.assertRendered(
    Marko._.rendered,
    4,
    Marko._.renderTemplate(__marko_internal_tag_4)()()({
      value: { a: 1, b: 2 } as const,
    }),
  );
  const record = Marko._.rendered.returns[4].value;
  Marko._.forInTag(
    {
      /*for*/ in: record,
    },
    (key, value) => {
      key;
      value;
      return Marko._.voidReturn;
    },
  );
  Marko._.forInTag(
    {
      /*for*/ in: record,
      by: (value, key) => `${value}-${key}`,
    },
    () =>
      //  ^?     ^?

      {
        return Marko._.voidReturn;
      },
  );
  Marko._.assertRendered(
    Marko._.rendered,
    5,
    Marko._.forInTag(
      {
        /*for*/ in: record,
      },
      (
        //                 ^?     ^?
        key,
      ) => {
        const __marko_internal_tag_5 = Marko._.resolveTemplate(
          import("../../../components/const/index.marko"),
        );
        Marko._.assertRendered(
          Marko._.rendered,
          6,
          Marko._.renderTemplate(__marko_internal_tag_5)()()({
            value: key,
          }),
        );
        const hoistedFromForIn = Marko._.rendered.returns[6].value;
        return new (class MarkoReturn<Return = void> {
          [Marko._.scope] = { hoistedFromForIn };
          declare return: Return;
          constructor(_?: Return) {}
        })();
      },
    ),
  );
  const __marko_internal_tag_6 = Marko._.interpolated`effect`;
  Marko._.renderDynamicTag(__marko_internal_tag_6)()()({
    value() {
      hoistedFromForIn;
      //^?
    },
  });
  Marko._.forToTag(
    {
      /*for*/ to: 10,
    },
    (index) => {
      index;
      return Marko._.voidReturn;
    },
  );
  Marko._.forToTag(
    {
      /*for*/ to: 10,
      by: (index) => `${index}`,
    },
    () =>
      //  ^?

      {
        return Marko._.voidReturn;
      },
  );
  Marko._.forToTag(
    {
      /*for*/ from: 1,
      to: 10,
    },
    (
      //             ^?
      index,
    ) => {
      index;
      return Marko._.voidReturn;
    },
  );
  Marko._.forToTag(
    {
      /*for*/ to: 10,
      step: 2,
    },
    (
      //  ^?
      index,
    ) => {
      index;
      return Marko._.voidReturn;
    },
  );
  Marko._.assertRendered(
    Marko._.rendered,
    7,
    Marko._.forToTag(
      {
        /*for*/ to: 10,
      },
      (
        //  ^?
        index,
      ) => {
        const __marko_internal_tag_7 = Marko._.resolveTemplate(
          import("../../../components/const/index.marko"),
        );
        Marko._.assertRendered(
          Marko._.rendered,
          8,
          Marko._.renderTemplate(__marko_internal_tag_7)()()({
            value: index,
          }),
        );
        const hoistedFromForTo = Marko._.rendered.returns[8].value;
        return new (class MarkoReturn<Return = void> {
          [Marko._.scope] = { hoistedFromForTo };
          declare return: Return;
          constructor(_?: Return) {}
        })();
      },
    ),
  );
  const __marko_internal_tag_8 = Marko._.interpolated`effect`;
  Marko._.renderDynamicTag(__marko_internal_tag_8)()()({
    value() {
      hoistedFromForTo;
      //^?
    },
  });
  Marko._.forTag(
    {
      /*for*/
    },
    (index) => {
      return Marko._.voidReturn;
    },
  );
  const { hoistedFromForOf, hoistedFromForIn, hoistedFromForTo } =
    Marko._.readScopes(Marko._.rendered);
  Marko._.noop({ hoistedFromForOf, hoistedFromForIn, hoistedFromForTo });
  return;
})();
export default new (class Template extends Marko._.Template<{
  render(
    input: Marko.TemplateInput<Input>,
    stream?: {
      write: (chunk: string) => void;
      end: (chunk?: string) => void;
    },
  ): Marko.Out<Component>;

  render(
    input: Marko.TemplateInput<Input>,
    cb?: (err: Error | null, result: Marko.RenderResult<Component>) => void,
  ): Marko.Out<Component>;

  renderSync(input: Marko.TemplateInput<Input>): Marko.RenderResult<Component>;

  renderToString(input: Marko.TemplateInput<Input>): string;

  stream(
    input: Marko.TemplateInput<Input>,
  ): ReadableStream<string> & NodeJS.ReadableStream;

  api: "class";
  _(): () => <__marko_internal_input extends unknown>(
    input: Marko.Directives &
      Input &
      Marko._.Relate<__marko_internal_input, Marko.Directives & Input>,
  ) => Marko._.ReturnWithScope<__marko_internal_input, void>;
}> {})();
