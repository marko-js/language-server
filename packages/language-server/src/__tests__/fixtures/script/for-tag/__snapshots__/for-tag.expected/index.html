<div>
  Repeated!
</div><div>
  placeholder placeholder placeholder

</div><div>

</div><div>
  
</div><div>
  placeholder placeholder

</div><div>

</div><div>
  
</div><div>
  placeholder

</div><div>

</div><div>
  placeholder

</div><div>
  placeholder

</div><div>
  
</div><div>
  Should error
</div>